File: lang/go/test/fun/vac/data_structures/ArrayList/ArrayListTest.go
package main

import (
    "fun/vac/data_structures/ArrayList"
    "fun/vac/util/TestRunner"
)

func testArrayList() bool {
    size := 8192

    list := ArrayList.New(0)

    for i := 1; i <= size; i++ {
        list.Append(i)
    }

    list.Shrink()

    if list.Size() != size {
        return false
    }

    if list.Capacity() != size {
        return false
    }

    for i := 0; i < size; i++ {
        if list.Get(i) != i + 1 {
            return false
        }
    }

    for i := 0; i < size; i++ {
        list.Set(i, size - i)
    }

    for i := 0; i < size; i++ {
        if list.Get(i) != size - i {
            return false
        }
    }

    var x int
    var y int

    for i, j := 0, size - 1; i < j; i, j = i + 1, j - 1 {
        x = list.Get(i)
        y = list.Get(j)

        list.Remove(i)
        list.Insert(i, y)
        list.Remove(j)
        list.Insert(j, x)
    }

    for i := 0; i < size; i++ {
        if list.Get(i) != i + 1 {
            return false
        }
    }

    for i := size; i >= 1; i-- {
        if list.Back() != i {
            return false
        }
        list.Eject()
    }

    list.Shrink()

    if !list.IsEmpty() {
        return false
    }

    return list.Capacity() == 0
}

func main() {
    TestRunner.ParseTest(testArrayList())
}

File: lang/ruby/test/vac.fun/data_structures/ArrayListTest.rb
require "vac.fun/data_structures/ArrayList"
require "vac.fun/util/TestRunner"

def testArrayList()
    size = 8192

    list = ArrayList::ArrayList.new(0)

    1.upto(size) do |i|
        list.append(i)
    end

    list.shrink()

    if list.size() != size
        return false
    end

    if list.capacity() != size
        return false
    end

    0.upto(size - 1) do |i|
        if list.get(i) != i + 1
            return false
        end
    end

    0.upto(size - 1) do |i|
        list.set(i, size - i)
    end

    0.upto(size - 1) do |i|
        if list.get(i) != size - i
            return false
        end
    end

    x = 0
    y = 0

    i, j = 0, size - 1
    while i < j
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1
    end

    0.upto(size - 1) do |i|
        if list.get(i) != i + 1
            return false
        end
    end

    size.downto(1) do |i|
        if list.back() != i
            return false
        end
        list.eject()
    end

    list.shrink()

    unless list.isEmpty()
        return false
    end

    list.capacity().zero?
end

if __FILE__ == $PROGRAM_NAME
    TestRunner.parseTest(testArrayList())
end